File: examples/java/ApibaraVehicleAuctionClient.java
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.stream.Collectors;

public class ApibaraVehicleAuctionClient {
    private final String apiKey;
    private final String baseUrl;
    private final HttpClient httpClient;

    public ApibaraVehicleAuctionClient(String apiKey) {
        this(apiKey, "https://apibara.tech/api/v1/vehicle-auction");
    }

    public ApibaraVehicleAuctionClient(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl;
        this.httpClient = HttpClient.newHttpClient();
    }

    public String get(String path, Map<String, String> params) throws IOException, InterruptedException {
        String query = params.entrySet().stream()
                .filter(e -> e.getValue() != null && !e.getValue().isBlank())
                .map(e -> encode(e.getKey()) + "=" + encode(e.getValue()))
                .collect(Collectors.joining("&"));

        String url = baseUrl + path + (query.isEmpty() ? "" : "?" + query);

        HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                .GET()
                .header("X-API-Key", apiKey)
                .header("Accept", "application/json")
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("Apibara API error " + response.statusCode() + ": " + response.body());
        }

        return response.body();
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    public static void main(String[] args) throws Exception {
        String apiKey = System.getenv("APIBARA_API_KEY");
        ApibaraVehicleAuctionClient client = new ApibaraVehicleAuctionClient(apiKey);

        String json = client.get("/vehicles", Map.of(
                "make", "Toyota",
                "model", "Camry",
                "year_from", "2018",
                "lot_sub_status", "Open",
                "per_page", "20"
        ));

        System.out.println(json);
    }
}

File: examples/kotlin/ApibaraVehicleAuctionClient.kt
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets

class ApibaraVehicleAuctionClient(
    private val apiKey: String,
    private val baseUrl: String = "https://apibara.tech/api/v1/vehicle-auction"
) {
    private val httpClient = HttpClient.newHttpClient()

    fun get(path: String, params: Map<String, String> = emptyMap()): String {
        val query = params
            .filterValues { it.isNotBlank() }
            .map { (key, value) -> "${encode(key)}=${encode(value)}" }
            .joinToString("&")

        val url = baseUrl + path + if (query.isBlank()) "" else "?$query"

        val request = HttpRequest.newBuilder(URI.create(url))
            .GET()
            .header("X-API-Key", apiKey)
            .header("Accept", "application/json")
            .build()

        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())

        if (response.statusCode() !in 200..299) {
            error("Apibara API error ${response.statusCode()}: ${response.body()}")
        }

        return response.body()
    }

    private fun encode(value: String): String = URLEncoder.encode(value, StandardCharsets.UTF_8)
}

fun main() {
    val apiKey = System.getenv("APIBARA_API_KEY") ?: error("Missing APIBARA_API_KEY")
    val client = ApibaraVehicleAuctionClient(apiKey)

    val vehicles = client.get("/vehicles", mapOf(
        "make" to "Toyota",
        "model" to "Camry",
        "year_from" to "2018",
        "lot_sub_status" to "Open",
        "per_page" to "20"
    ))

    println(vehicles)
}
